File: javascript/json-transform/src/functions/TransformerFunctionXmlParse.ts
import { X2jOptions, XMLParser } from "fast-xml-parser";
import TransformerFunction from "./common/TransformerFunction";
import FunctionContext from "./common/FunctionContext";
import { ArgType } from "./common/ArgType";

class TransformerFunctionXmlParse extends TransformerFunction {
  constructor() {
    super({
      argsSet: [
        { name: "keep_strings", type: ArgType.Boolean, defaultValue: false },
        { name: "attr_prefix", type: ArgType.String, defaultValue: "@" },
        { name: "cdata_prop_name", type: ArgType.String, defaultValue: "#text" },
        { name: "array_tags", type: ArgType.Array },
      ],
    });
  }

  override async apply(context: FunctionContext): Promise<any> {
    const xml = await context.getString(null);
    if (xml == null) return null;
    try {
      const keepStrings = (await context.getBoolean("keep_strings")) ?? undefined;
      const attrPrefix = (await context.getString("attr_prefix")) ?? undefined;
      const cDataTagName = (await context.getString("cdata_prop_name")) ?? undefined;
      const arrayTags = (await context.getJsonArray("array_tags")) ?? undefined;

      const options = {
        attributeNamePrefix: attrPrefix,
        allowBooleanAttributes: true,
        ignoreAttributes: false,
        textNodeName: cDataTagName,
        parseTagValue: !keepStrings,
        parseAttributeValue: !keepStrings,
        maxNestedTags: 256,
      } as X2jOptions;
      if (arrayTags?.length) {
        const tagsSet = new Set(arrayTags);
        options.isArray = x => tagsSet.has(x);
      }
      // if (!keepStrings) {
      //   options.attributeValueProcessor = (_attrName, attrValue) => {
      //     console.log("attrValue", attrValue);
      //     if (attrValue === "true") return true;
      //     if (attrValue === "false") return false;
      //     return attrValue;
      //   };
      // }

      return new XMLParser(options).parse(xml);
    } catch (e: any) {
      console.warn(context.getAlias() + " failed", e);
      return null;
    }
  }
}

export default TransformerFunctionXmlParse;

File: java/json-transform/src/main/java/ai/bizone/jsontransform/functions/TransformerFunctionXmlParse.java
package ai.bizone.jsontransform.functions;

import ai.bizone.jsontransform.functions.common.*;
import ai.bizone.jsontransform.functions.common.*;
import ai.bizone.jsontransform.formats.xml.XmlFormat;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.Map;
import java.util.stream.Collectors;

public class TransformerFunctionXmlParse extends TransformerFunction {
    static final Logger logger = LoggerFactory.getLogger(TransformerFunctionXmlParse.class);
    public TransformerFunctionXmlParse() {
        super(FunctionDescription.of(
            Map.of(
            "keep_strings", ArgumentType.of(ArgType.Boolean).position(0).defaultValue(false),
            "attr_prefix", ArgumentType.of(ArgType.String).position(1).defaultValue("@"),
            "cdata_prop_name", ArgumentType.of(ArgType.String).position(2).defaultValue("#text"),
            "array_tags", ArgumentType.of(ArgType.Array).position(3)
            )
        ));
    }
    @Override
    public Object apply(FunctionContext context) {
        var xml = context.getString(null);
        if (xml == null)
            return null;
        try {
            var keepStrings = context.getBoolean("keep_strings");
            var attrPrefix = context.getString("attr_prefix");
            var cDataPropName = context.getString("cdata_prop_name");
            var arrayTags = context.getJsonArray("array_tags");
            var adapter = context.getAdapter();
            return new XmlFormat(adapter,
                null,
                null,
                null,
                attrPrefix,
                cDataPropName,
                keepStrings,
                arrayTags == null ? null : adapter.stream(arrayTags).map(context::getAsString).collect(Collectors.toSet())
            ).deserialize(xml);
        } catch (Exception e) {
            logger.warn(context.getAlias() + " failed", e);
            return null;
        }
    }
}
